From 5aa14045ce33b258e02223a7b6e153df8c8e1457 Mon Sep 17 00:00:00 2001
From: Yu Kuai <yukuai3@huawei.com>
Date: Mon, 8 May 2023 17:41:37 +0800
Subject: [PATCH 1620/1646] [Huawei] block: shutdown blktrace in
 blk_release_queue()

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: 188711

----------------------------------------

Commit 99d055b4fd4b ("block: remove per-disk debugfs files in
blk_unregister_queue") move blk_trace_shutdown() from
blk_release_queue() to blk_unregister_queue(). However, blktrace can
still be enabled through ioctl after blk_unregister_queue(), and
blktrace will be leaked in this case.

Fix the problem by calling blk_trace_shutdown() in blk_release_queue().

Signed-off-by: Yu Kuai <yukuai3@huawei.com>
---
 block/blk-sysfs.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/block/blk-sysfs.c b/block/blk-sysfs.c
index 4566f01f8427..8d696d586731 100644
--- a/block/blk-sysfs.c
+++ b/block/blk-sysfs.c
@@ -801,6 +801,10 @@ static void blk_release_queue(struct kobject *kobj)
 	if (queue_is_mq(q))
 		blk_mq_release(q);
 
+	mutex_lock(&q->debugfs_mutex);
+	blk_trace_shutdown(q);
+	mutex_unlock(&q->debugfs_mutex);
+
 	bioset_exit(&q->bio_split);
 
 	ida_simple_remove(&blk_queue_ida, q->id);
-- 
2.17.1

